master/buildbot/worker/suspendable.py

```diff
@@ -0,0 +1,364 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Portions Copyright Buildbot Team Members
+
+import os
+import stat
+
+from twisted.internet import defer
+from twisted.internet import utils
+from twisted.python import log
+from zope.interface import implementer
+
+from buildbot import config
+from buildbot import interfaces
+from buildbot.util import Notifier
+from buildbot.util import misc
+from buildbot.util import private_tempdir
+from buildbot.util import service
+from buildbot.util.git import getSshArgsForKeys
+from buildbot.util.git import getSshKnownHostsContents
+from buildbot.worker import AbstractWorker
+from buildbot.worker.latent import AbstractLatentWorker
+
+
+@implementer(interfaces.ISuspendableWorker)
+class SuspendableWorker(AbstractLatentWorker):
```

Comment:
I think this logic should be part of the base LatentWorker.

We should be able to attach a machine_manager to all the latent workers.

For example this can be something we use for Kubernetes cluster.

We could use a machine manager to completely shutdown the cluster, and avoid costs during long interruption times.


Same for docker latent worker, one could start an EC2 machine for 4 docker latent worker.

All those use cases for me make sense for the wide range of user while your usecase with ssh and WOL, I am not very convinced will have more than one user (no offense)

Reply:
No offence taken :) Your suggestions make sense, will implement them. What do you think about the possibility of having machine locks? If that makes sense, I would think about it while reworking the PR so that future expansion is easier.

Reply:
> machine locks

why not, but lets implement it with baby steps

Reply:
Yes, sure.